File: ximple-dev_cpp/vtd-xml/unionExpr.cpp
/* 
* Copyright (C) 2002-2010 XimpleWare, info@ximpleware.com
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/
#include "unionExpr.h"
#pragma warning(disable:4355)
using namespace com_ximpleware;

UnionExpr::UnionExpr(Expr *e):
current(this),
next(NULL),
fe(e),
evalState(0),
ih(NULL)
{
}
UnionExpr::~UnionExpr(){
	UnionExpr *tmp, *tmp2;
	delete(fe);
	delete(ih);
	tmp = next;

	while(tmp!= NULL){
		delete(tmp->fe);
		delete(tmp->ih);
		tmp2 = tmp;
		tmp = tmp->next;		
		delete(tmp2);
	}
}

bool UnionExpr::evalBoolean(VTDNav *vn){
	
	bool b = false;
	int size;
	if (fe->isNodeSet()==false){
		return fe->evalBoolean(vn);
	}else{
			vn->push2();
			/* record teh stack size*/
			size = vn->contextBuf2->size;
			try{	
				b = (evalNodeSet(vn) != -1);
			}catch (...){
			}
			/*rewind stack */
			vn->contextBuf2->size = size;
			reset(vn);
			vn->pop2();
			return b;
	}
}

double UnionExpr::evalNumber(VTDNav *vn){
	double d = 0.0;
	int a;
	if (fe->isNodeSet()==false){   
		return fe->evalNumber(vn);   
	}
	a = getStringIndex(vn);
	try{
		if (a!=-1) return vn->parseDouble(a);
	}catch (...){
	}
	return 0/d;
}

int UnionExpr::evalNodeSet(VTDNav *vn){
	int a;
	if (next == NULL) {
		return fe->evalNodeSet(vn);
	} else {
		while (true) {
			switch (evalState) {
				case 0:
					if (ih == NULL ){
						//exception ee;
						try{
							ih = new IntHash();
						}
						catch(...){
							throw;
						}
					}
					if (current != NULL) {
						vn->push2();
						while ((a = current->fe->evalNodeSet(vn))!= -1) {
								if (isUnique_une(a)) {
									evalState = 1;
									return a;
								}
						}
						evalState = 2;
						vn->pop2();
						break;
					} else
						evalState = 3;
					break;

				case 1:
					while ((a = current->fe->evalNodeSet(vn)) != -1) 
					{
						if (isUnique_une(a)) {
							evalState = 1;
							return a;
						}
					}
					evalState = 2;
					vn->pop2();
					break;

				case 2:
					current = current->next;
					if (current != NULL) {
						vn->push2();
						while ((a = current->fe->evalNodeSet(vn)) != -1) {
								if (isUnique_une(a)) {
									evalState = 1;
									return a;
								}
						}
						vn->pop2();
						break;
					} else
						evalState = 3;
					break;

				case 3:
					return -1;

				default:
					throw OtherException("Invalid state evaluating unionExpr");
			}
		}
	}
}

UCSChar* UnionExpr::evalString(VTDNav *vn){
	int a;
	if (fe->isNodeSet()==false){   
		return fe->evalString(vn);   
	}
	a = getStringIndex(vn);
	try {
		if (a != -1)
			return vn->toString(a);
	} catch (std::bad_alloc& ) {
		throw;
	}
	return createEmptyString();	
}

void UnionExpr::reset(VTDNav *vn){
	UnionExpr *tmp;
	fe->reset(vn);
	current = this;
	tmp = next;
    while (tmp != NULL) {
        tmp->fe->reset(vn);
        tmp = tmp->next;
    }
	if (ih!=NULL)
		ih->reset();
	evalState = 0;
}

void UnionExpr::toString(UCSChar *string){
       if (next == NULL)
            fe->toString(string);
	   else{
            fe->toString(string);
			wprintf(L" | ");
			next->toString(string);
	   }
}

bool UnionExpr::isNumerical(){
	return fe->isNumerical();
}
bool UnionExpr::isNodeSet(){
	return fe->isNodeSet();
}
bool UnionExpr::isString(){
	return fe->isString();
}
bool UnionExpr::isBoolean(){
	return fe->isBoolean();
}

bool UnionExpr::requireContextSize(){
// boolean b = false;
	UnionExpr* tmp = this;

	while (tmp != NULL) {
		if (tmp->fe->requireContextSize()== true){
			return true;
		}            
		tmp = tmp->next;
	}        
	return false;
}
void UnionExpr::setContextSize(int size){
	UnionExpr *tmp;
    current = this;
    current->fe->setContextSize( size);
    tmp = next;
    while (tmp != NULL) {
        tmp->fe->setContextSize( size);
        tmp = tmp->next;
    }
}

void UnionExpr::setPosition(int pos){
	UnionExpr *tmp;
    current = this;
    current->fe->setPosition( pos);
    tmp = next;
    while (tmp != NULL) {
        tmp->fe->setPosition(pos);
        tmp = tmp->next;
    }
}

int UnionExpr::adjust(int n){
	int i=fe->adjust(n);
	UnionExpr *tmp = NULL;		
	tmp = next;
	while (tmp != NULL) {
		tmp->fe->adjust(n);
		tmp = tmp->next;
	}
	if (ih!=NULL && i==ih->e)
	{}else{
		delete (ih);
		ih = new IntHash(i);
	}
	return i;
}

File: ximple-dev_csharp/ximple-dev_csharp/com/ximpleware/UnionExpr.cs
/* 
* Copyright (C) 2002-2011 XimpleWare, info@ximpleware.com
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/
using System;
//UPGRADE_TODO: The 'com.ximpleware.xpath' package was found, but not necessarily converted. References to it may cause compile errors. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1280'"
using com.ximpleware.xpath;
namespace com.ximpleware
{
	
	public class UnionExpr:Expr
	{
		internal intHash ih;
		
		internal Expr e;
		
		internal UnionExpr next;
		
		internal UnionExpr current;
		
		internal int state;

        public override int adjust(int n)
        {
            int i = e.adjust(n);
            if (ih != null && i == ih.e)
            { }
            else
                ih = new intHash(i);
            UnionExpr tmp = this.next;
            while (tmp != null)
            {
                tmp.e.adjust(n);
                tmp = tmp.next;
            }
            return i;
        }
		public UnionExpr(Expr e1)
		{
			e = e1;
			next = null;
			current = this;
			ih = null;
			state = 0;
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#evalBoolean(com.ximpleware.VTDNav)
		*/
		public override bool evalBoolean(VTDNav vn)
		{
            if (e.NodeSet==false)
            {
                return e.evalBoolean(vn);
            }
            else
            {
                bool a = false;
                vn.push2();
                // record teh stack size
                int size = vn.contextStack2.size;
                try
                {
                    a = (evalNodeSet(vn) != -1);
                }
                catch (System.Exception ee)
                {
                }
                //rewind stack
                vn.contextStack2.size = size;
                reset(vn);
                vn.pop2();
                return a;
            }
            /*if (e.Numerical)
            {
                double dval = e.evalNumber(vn);

                if (dval == 0.0 || System.Double.IsNaN(dval))
                    return false;
                return true;
            }

            String s = e.evalString(vn);
            if (s == null || s.Length == 0)
                return false;
            return true;*/

		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#evalNumber(com.ximpleware.VTDNav)
		*/
		public override double evalNumber(VTDNav vn)
		{
            if (e.NodeSet==false)
                return e.evalNumber(vn);
            int a = getStringIndex(vn);
            try
            {
                if (a != -1)
                    return vn.parseDouble(a);
            }
            catch (NavException ee)
            {
            }
            return Double.NaN;
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#evalNodeSet(com.ximpleware.VTDNav)
		*/
		public override int evalNodeSet(VTDNav vn)
		{
			int a;
			if (this.next == null)
			{
				return e.evalNodeSet(vn);
			}
			else
			{
				while (true)
				{
					switch (state)
					{
						
						case 0: 
							if (ih == null)
							  ih = new intHash();
							if (current != null)
							{
								vn.push2();
								while ((a = current.e.evalNodeSet(vn)) != - 1)
								{
									if (isUnique(a))
									{
										state = 1;
										return a;
									}
								}
								state = 2;
								vn.pop2();
								break;
							}
							else
								state = 3;
							break;
						
						
						case 1: 
							while ((a = current.e.evalNodeSet(vn)) != - 1)
							{
								if (isUnique(a))
								{
									state = 1;
									return a;
								}
							}
							state = 2;
							vn.pop2();
							break;
						
						
						case 2: 
							current = current.next;
							if (current != null)
							{
								vn.push2();
								while ((a = current.e.evalNodeSet(vn)) != - 1)
								{
									if (isUnique(a))
									{
										state = 1;
										return a;
									}
								}
								vn.pop2();
								break;
							}
							else
								state = 3;
							break;
						
						
						case 3: 
							return - 1;
						
						
						default: 
							throw new XPathEvalException("Invalid state evaluating PathExpr");
						
					}
				}
			}
			
			/*
			* default: throw new XPathEvalException( "Invalid state evaluating
			* PathExpr");
			*/
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#evalString(com.ximpleware.VTDNav)
		*/
		public override System.String evalString(VTDNav vn)
		{
              if (e.NodeSet == false){
                return e.evalString(vn);                
              }
              int a = getStringIndex(vn);
              try
              {
                  if (a != -1)
                      return vn.toString(a);
              }
              catch (NavException ee)
              {
              }
              return "";
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#reset(com.ximpleware.VTDNav)
		*/
		public override void  reset(VTDNav vn)
		{
			// travese el list and reset every expression
			e.reset(vn);
			current = this;
			UnionExpr tmp = this.next;
			while (tmp != null)
			{
				tmp.e.reset(vn);
				tmp = tmp.next;
			}
			if (ih != null)
				ih.reset();
			state = 0;
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see java.lang.Object#toString()
		*/
		public override System.String ToString()
		{
			// TODO Auto-generated method stub
			
			if (this.next == null)
			{
				return this.e.ToString();
			}
			else
			{
				return this.e.ToString() + " | " + this.next.ToString();
			}
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#isNumerical()
		*/
		public override bool Numerical
		{
            get
            {
                // TODO Auto-generated method stub
                return e.Numerical;
            }
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#isNodeSet()
		*/
		public override bool NodeSet
		{
            get
            {
                // TODO Auto-generated method stub
                return e.NodeSet;
            }
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#isString()
		*/
		public override bool String
		{
            get
            {
                // TODO Auto-generated method stub
                return e.String;
            }
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#isBoolean()
		*/
		public override bool Boolean
		{
            get
            {
                // TODO Auto-generated method stub
                return false;
            }
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#requireContextSize()
		*/
		public override bool requireContextSize()
		{
            //bool b = false;
            UnionExpr tmp = this;
            while (tmp != null)
            {
                if (tmp.e.requireContextSize() == true)
                {
                    return true;
                }
                tmp = tmp.next;
            }
            return false;
		}
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#setContextSize(int)
		*/
        override public int ContextSize
        {
            set
            {
                current = this;
                current.e.ContextSize = value;
                UnionExpr tmp = this.next;
                while (tmp != null)
                {
                    tmp.e.ContextSize = value;
                    tmp = tmp.next;
                }
            }

        }
		
		/*
		* (non-Javadoc)
		* 
		* @see com.ximpleware.xpath.Expr#setPosition(int)
		*/
        override public int Position
        {
            set
            {
                current = this;
                current.e.Position = value;
                UnionExpr tmp = this.next;
                while (tmp != null)
                {
                    tmp.e.Position = value;
                    tmp = tmp.next;
                }

            }

        }
		
		public virtual bool isUnique(int i)
		{
			return ih.isUnique(i);
		}
	}
}
